Unlock content immediately for posts with the subscriber access level

romarioraffington opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Closes  https://github.com/Automattic/wp-calypso/issues/69664
Closes https://github.com/Automattic/wp-calypso/issues/70119

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
We do not immediately unlock a post with the subscriber access level. Instead, we send a verification email. This is not an ideal user experience. This PR gives immediate access to subscribers after subscribing once the post has the subscriber access level.

This PR and patch will also hide the free option in the `screen-choose-plan.php` template if the post is marked for paid subscribers.


#### Testing instructions:
See D92672-code

romarioraffington committed:
Unlock immediately for subscribers

When a user subscribes to a post with an access level of 'subscribers' allow the user to view the post immediately

romarioraffington committed:
Unlock content for Jetpack sites subscribers and update const

artpi reviewed (changes requested):
I see where this approach is going.

You tapped into the return of the current subscribe block flow to capture these parameters and unlock the content.
This will work short-term, but won't work with setting up the cookie for when somebody navigates from the site and back.

The flow would work like so:

1. Click on subscribe
2. Modal opens
3. Choose free option
4. Modal closes
5. The old subscribe form submits
6. Redirects to WPCOM
7. Subscribes user, sends confirmation email
8. Redirects back with query string
9. This PR is intercepting this query string and unlocking content based on it's contents.

What I think should be happening:

1. Click on subscribe
2. Modal opens
3. Choose free option
4. This PR should change how this free option behaves. It should subscribe user inside the modal and redirect with a valid JWT token (with blog_sub=confirming) that in turn will unlock content

artpi commented:
I'm open to discussion!

romarioraffington committed:
[not verified] No longer need to listen to jetpack_subscribe_continue

Subscription for free plans will be handle via the subscribe modal

romarioraffington committed:
No longer need to listen to jetpack_subscribe_continue

Subscription for free plans will be handle via the subscribe modal

romarioraffington committed:
Rely on token to unlock content for free subscribers

millerf commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -71,8 +74,21 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
 		if ( $is_valid_token ) {
-			if ( 'subscribers' === $access_level ) {
-				return 'active' === $payload['blog_sub'];
+			if ( self::POST_ACCESS_LEVEL_SUBSCRIBERS === $access_level ) {
```
I have modified all that logic in [this PR](https://github.com/Automattic/jetpack/pull/27352). Do you mind to wait and rebase on it? It should be minimal work compared to the other way around ....

millerf reviewed (approved):
LGTM- added a comment about merging

romarioraffington committed:
Pass the post_access_level to WPCOM form

This will allow us to customize the options we show in the screen-choose-plan.php template based on the post access level

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -71,8 +74,21 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+				 * Active: user has a valid subscription
+				 * Pending: user has subscribed for a free plan
+				 */
+				return in_array(
```
Yes, that is what I was thinking about!

romarioraffington committed:
Correct comment

romarioraffington committed:
Add a default param to postMeta (#27508)

useEntityProp can return undefined. To prevent errors add a default state

romarioraffington committed:
Merge branch 'paid-newsletters' into add/unlock-content-immediately-for-subscribers
